src/core/call/request_buffer.cc

```diff
@@ -0,0 +1,162 @@
+// Copyright 2024 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/core/call/request_buffer.h"
+
+#include <cstdint>
+
+#include "absl/types/optional.h"
+
+namespace grpc_core {
+
+StatusFlag RequestBuffer::PushClientInitialMetadata(ClientMetadataHandle md) {
+  MutexLock lock(&mu_);
+  if (absl::get_if<Cancelled>(&state_)) return Failure{};
+  auto& buffering = absl::get<Buffering>(state_);
+  CHECK_EQ(buffering.initial_metadata.get(), nullptr);
+  buffering.initial_metadata = std::move(md);
+  buffering.buffered += buffering.initial_metadata->TransportSize();
+  WakeupAsyncAllPullers();
+  return Success{};
+}
+
+Poll<ValueOrFailure<size_t>> RequestBuffer::PollPushMessage(
+    MessageHandle& message) {
+  MutexLock lock(&mu_);
+  if (absl::get_if<Cancelled>(&state_)) return Failure{};
+  size_t buffered = 0;
+  if (auto* buffering = absl::get_if<Buffering>(&state_)) {
+    buffering->buffered += message->payload()->Length();
+    if (winner_ == nullptr) buffered = buffering->buffered;
+    buffering->messages.push_back(std::move(message));
+  } else {
+    auto& streaming = absl::get<Streaming>(state_);
+    CHECK_EQ(streaming.end_of_stream, false);
+    if (streaming.message != nullptr) {
+      return PendingPush();
+    }
+    streaming.message = std::move(message);
+  }
+  WakeupAsyncAllPullers();
+  return buffered;
+}
+
+StatusFlag RequestBuffer::FinishSends() {
+  MutexLock lock(&mu_);
+  if (absl::get_if<Cancelled>(&state_)) return Failure{};
+  if (auto* buffering = absl::get_if<Buffering>(&state_)) {
+    Buffered buffered(std::move(buffering->initial_metadata),
+                      std::move(buffering->messages));
+    state_.emplace<Buffered>(std::move(buffered));
+  } else {
+    auto& streaming = absl::get<Streaming>(state_);
+    CHECK_EQ(streaming.end_of_stream, false);
+    streaming.end_of_stream = true;
+  }
+  WakeupAsyncAllPullers();
+  return Success{};
+}
+
+void RequestBuffer::Cancel(absl::Status error) {
+  MutexLock lock(&mu_);
+  if (absl::holds_alternative<Cancelled>(state_)) return;
+  state_.emplace<Cancelled>(std::move(error));
+  WakeupAsyncAllPullers();
+}
+
+void RequestBuffer::SwitchToStreaming(Reader* winner) {
+  MutexLock lock(&mu_);
+  CHECK_EQ(winner_, nullptr);
+  winner_ = winner;
+  if (auto* buffering = absl::get_if<Buffering>(&state_)) {
+    if (buffering->initial_metadata != nullptr &&
```

Comment:
Since we're switching to `Streaming` state only if the winning reader has already pulled all of the data, does that mean that we expect the caller to call this method only after the winning reader has pulled all the data?  What happens if the caller calls this method too early?

Reply:
It's ok to call this method early -- the winning reader will read to the end of the buffered messages and then we'll switch to streaming mode.

Suppose you have a hedged request and one attempt fails in a retryable fashion - we'd likely want to be able to have the other continue even if it hasn't yet sent all the messages.